fontir/src/ir.rs

```diff
@@ -146,23 +146,103 @@ pub struct GlobalMetrics(
+}
+
+/// Adjust Y offset based on italic angle, to get X offset.
+///
+///  <https://github.com/googlefonts/ufo2ft/blob/0d2688cd847d003b41104534d16973f72ef26c40/Lib/ufo2ft/outlineCompiler.py#L571-L573>
+fn adjust_offset(offset: f32, angle: f32) -> f32 {
+    if angle.abs() >= f32::EPSILON {
+        offset * (-angle).to_radians().tan()
+    } else {
+        0.0
+    }
 }
 
 impl GlobalMetrics {
     /// Creates a new [GlobalMetrics] with default values at the default location.
-    pub fn new(default_location: NormalizedLocation, units_per_em: u16) -> GlobalMetrics {
+    pub fn new(
+        default_location: NormalizedLocation,
+        units_per_em: u16,
+        x_height: Option<f32>,
+    ) -> GlobalMetrics {
         let mut metrics = GlobalMetrics(HashMap::new());
         let mut set = |metric, value| metrics.set(metric, default_location.clone(), value);
 
         // https://github.com/googlefonts/ufo2ft/blob/fca66fe3ea1ea88ffb36f8264b21ce042d3afd05/Lib/ufo2ft/fontInfoData.py#L38-L45
-        set(GlobalMetric::Ascender, 0.8 * units_per_em as f32);
-        set(GlobalMetric::Descender, -0.2 * units_per_em as f32);
+        let ascender = 0.8 * units_per_em as f32;
+        let descender = -0.2 * units_per_em as f32;
+        set(GlobalMetric::Ascender, ascender);
+        set(GlobalMetric::Descender, descender);
+
+        // https://github.com/googlefonts/ufo2ft/blob/0d2688cd847d003b41104534d16973f72ef26c40/Lib/ufo2ft/fontInfoData.py#L229-L238
+        let typo_line_gap = units_per_em as f32 * 1.2 + descender - ascender;
+        let typo_line_gap = if typo_line_gap > 0.0 {
+            typo_line_gap
+        } else {
+            0.0
+        };
+        set(GlobalMetric::Os2TypoLineGap, typo_line_gap);
+
+        // https://github.com/googlefonts/ufo2ft/blob/0d2688cd847d003b41104534d16973f72ef26c40/Lib/ufo2ft/fontInfoData.py#L215-L226
+        set(GlobalMetric::Os2TypoAscender, ascender + typo_line_gap);
+        set(GlobalMetric::Os2TypoDescender, descender);
+
+        // https://github.com/googlefonts/ufo2ft/blob/0d2688cd847d003b41104534d16973f72ef26c40/Lib/ufo2ft/fontInfoData.py#L241-L254
+        set(GlobalMetric::Os2WinAscent, ascender);
+        set(GlobalMetric::Os2WinDescent, descender);
 
         // https://github.com/googlefonts/ufo2ft/blob/fca66fe3ea1ea88ffb36f8264b21ce042d3afd05/Lib/ufo2ft/fontInfoData.py#L48-L55
         set(GlobalMetric::CapHeight, 0.7 * units_per_em as f32);
-        set(GlobalMetric::XHeight, 0.5 * units_per_em as f32);
+        let x_height = x_height.unwrap_or(0.5 * units_per_em as f32);
+        set(GlobalMetric::XHeight, x_height);
+
+        // https://github.com/googlefonts/ufo2ft/blob/0d2688cd847d003b41104534d16973f72ef26c40/Lib/ufo2ft/fontInfoData.py#L360
+        let italic_angle = 0.0;
+        set(GlobalMetric::ItalicAngle, italic_angle);
+
+        // https://github.com/googlefonts/ufo2ft/blob/0d2688cd847d003b41104534d16973f72ef26c40/Lib/ufo2ft/fontInfoData.py#L360
```

Comment:
the link is the same as the one before for italiAngle, copypasta?

Reply:
That was exceptionally wrong, good catch. Fixed.